contracts/nois-gateway/src/contract.rs

```diff
@@ -110,14 +183,60 @@ pub fn ibc_channel_open(
 
 #[entry_point]
 pub fn ibc_channel_connect(
-    _deps: DepsMut,
-    _env: Env,
+    deps: DepsMut,
+    env: Env,
     msg: IbcChannelConnectMsg,
-) -> StdResult<IbcBasicResponse> {
-    let channel = msg.channel();
-    let chan_id = &channel.endpoint.channel_id;
+) -> Result<IbcBasicResponse, ContractError> {
+    let channel = match msg {
+        IbcChannelConnectMsg::OpenAck { .. } => return Err(ContractError::MustBeChainB),
+        IbcChannelConnectMsg::OpenConfirm { channel, .. } => channel,
+    };
+    let chan_id = channel.endpoint.channel_id;
+
+    let config = CONFIG.load(deps.storage)?;
+
+    // Instantiate payment contract
+    let creator = deps.api.addr_canonicalize(env.contract.address.as_str())?;
+    let CodeInfoResponse { checksum, .. } =
+        deps.querier.query_wasm_code_info(config.payment_code_id)?;
+    let salt = hash_channel(&chan_id);
+    #[allow(unused)]
+    let address = instantiate2_address(&checksum, &creator, &salt)
+        .map_err(|e| StdError::generic_err(format!("Cound not generate address: {}", e)))?;
+
+    let address = {
+        #[cfg(not(test))]
+        {
+            deps.api.addr_humanize(&address)?
+        }
+        #[cfg(test)]
+        cosmwasm_std::Addr::unchecked("some payment address")
+    };
+    PAYMENT_ADDRESSES.save(deps.storage, &chan_id, &address)?;
+
+    let msg = WasmMsg::Instantiate2 {
+        admin: Some(env.contract.address.into()), // Only gateway can update the contracts it created
```

Comment:
why not use creator here?

Reply:
My idea was that the gateway maintains a list of instances anyways. So it can upgrade them all. This also simplifies security because we only need to worry about securing the gateway.

Reply:
Sorry what I meant here is more that there is in the code a bit above a variable called creator that gets the same value 
not sure how to get a url for a line in ggithub but here'e the line I'm referring to  
let creator = deps.api.addr_canonicalize(env.contract.address.as_str())?;

Reply:
Ah, right. `creator` the canonical address representation of `env.contract.address`. This is 32 raw bytes as Vec<u8>. Here we need a bech32 String, which is exactly what `env.contract.address` is already.

Reply:
Oh alright makes sense